ui/raidboss/emulator/data/network_log_converter/LineEvent.ts

```diff
@@ -61,3 +61,70 @@ export default class LineEvent {
       ).toString(16), 16);
   }
 }
+
+// Type guards for these interfaces require their own descriptor property
+// because we don't want every line event with an id/name
+// to update combatant state, for example
+export interface LineEventSource {
+  readonly isSource: true;
+  readonly id: string;
+  readonly name: string;
+  readonly x?: number;
+  readonly y?: number;
+  readonly z?: number;
+  readonly heading?: number;
+  readonly targetable?: boolean;
+  readonly hp?: number;
+  readonly maxHp?: number;
+  readonly mp?: number;
+  readonly maxMp?: number;
+}
+
+export const isLineEventSource =
+  (line: LineEventSource | LineEvent): line is LineEventSource => {
+    return 'isSource' in line;
+  };
+
+export interface LineEventTarget {
+  readonly isTarget: true;
+  readonly targetId: string;
+  readonly targetName: string;
+  readonly targetX?: number;
+  readonly targetY?: number;
+  readonly targetZ?: number;
+  readonly targetHeading?: number;
+  readonly targetHp?: number;
+  readonly targetMaxHp?: number;
+  readonly targetMp?: number;
+  readonly targetMaxMp?: number;
+}
+
+export const isLineEventTarget =
+  (line: LineEventTarget | LineEvent): line is LineEventTarget => {
+    return 'isTarget' in line;
+  };
+
+export interface LineEventJobLevel {
+  readonly isJobLevel: true;
+  readonly job?: string;
+  readonly jobId?: number;
+  readonly level?: number;
+}
+
+export const isLineEventJobLevel =
+  (line: LineEvent | LineEventSource | LineEventTarget | LineEventJobLevel):
+  line is LineEventJobLevel => {
+    return 'isJobLevel' in line;
+  };
+
+export interface LineEventAbility {
+  readonly isAbility: true;
+  readonly abilityId?: number;
+  readonly abilityName?: string;
+}
+
+export const isLineEventAbility =
+  (line: LineEvent | LineEventSource | LineEventTarget | LineEventAbility):
```

Comment:
This was a little confusing for me to read.  I understand why it needs to be there, but what do you think about just having `(line: LineEvent)` here instead of this confusing union and solving the Typescript error by having `LineEventAbility extends LineEvent`.  I think that would solve this, and might make `LineEventAbility` a more useful type at the same time.

Reply:
Resolved the one above and we can talk about it on this one.

If each of these extends LineEvent, then how can a LineEvent be both LineEventSource and LineEventTarget? Typescript doesn't support multiple inheritance for classes, only interfaces, and LineEvent is a class.

Reply:
I was trying to suggest that `LineEventSource extends LineEvent` and `LineEventTarget extends LineEvent`.  `LineEvent` itself just stays as a class.

Reply:
Except that this isn't valid syntax:

```typescript
class LineEvent {}

interface LineEventSource extends LineEvent {}
```

An interface can't extend a class.

If it were instead:

```typescript
class LineEvent {}

class LineEventSource extends LineEvent {}
```

Then this is no longer possible because a class can't extend multiple classes:

```typescript
class LineEvent0x1A extends LineEvent, LineEventSource, LineEventJobLevel {}
```

Reply:
https://www.typescriptlang.org/play?strict=true#code/MYGwhgzhAEAyCWA7ApgUQG7MQF2gbwF8BYAKFKW2QCcAzMYZOJNTHAZQHsBXKh6ZAB6VEAExgIUGLLkKlSQA seems to say it's ok? `tsc` also doesn't complain to me when I make that change in your patch.  Can you point me at the documentation for this for my own understanding?

Reply:
I found this page that states that it's possible, but it's marked as deprecated and the newer page doesn't state directly that it's okay?

https://www.typescriptlang.org/docs/handbook/interfaces.html#interfaces-extending-classes

Reply:
"it" here being deprecated is just the page as a whole, but not that feature specifically.  Agreed that the newer page doesn't state it, but I also can't find anything else about it either.  ("Type guards" aren't in the newer documentation either, but maybe I'm just missing them.)

I guess I don't feel that strongly about it.  It seems pretty reasonable to me, but if it seems sketchy to you, then this is fine as-is.

Reply:
Looks like type guards are in the documentation but under a different name?
https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

I'm fine with `interface extends class` but I'm always a bit leery of using something that might be deprecated.

I don't see anything in the typescript 2.0 release notes that would indicate that the syntax/feature would be deprecated, so maybe they just didn't bring it forward in their documentation?